src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+            base.VisitConvertedSwitchExpression(node);
+            return null;
+        }
+
+        public override BoundNode? VisitSwitchSection(BoundSwitchSection node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchSection(node);
+        }
+
+        public override BoundNode? VisitSwitchExpressionArm(BoundSwitchExpressionArm node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchExpressionArm(node);
+        }
+
+        public override BoundNode? VisitCatchBlock(BoundCatchBlock node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitCatchBlock(node);
+        }
+
+        public override BoundNode? VisitLocal(BoundLocal node)
+        {
+            Debug.Assert(_localEscapeScopes?.ContainsKey(node.LocalSymbol) == true ||
+                (node.LocalSymbol.ContainingSymbol is SynthesizedSimpleProgramEntryPointSymbol entryPoint && _symbol != entryPoint));
+
+            return base.VisitLocal(node);
+        }
+
+        private void AddLocalScopes(LocalSymbol local, uint refEscapeScope, uint valEscapeScope)
+        {
+            var scope = _useUpdatedEscapeRules ? local.Scope : DeclarationScope.Unscoped;
+            if (scope != DeclarationScope.Unscoped)
+            {
+                // From https://github.com/dotnet/csharplang/blob/main/csharp-11.0/proposals/low-level-struct-improvements.md:
+                //
+                // | Parameter or Local     | ref-safe-to-escape | safe-to-escape |
+                // |------------------------|--------------------|----------------|
+                // | Span<int> s            | current method     | calling method |
+                // | scoped Span<int> s     | current method     | current method |
+                // | ref Span<int> s        | calling method     | calling method |
+                // | scoped ref Span<int> s | current method     | calling method |
+
+                refEscapeScope = scope == DeclarationScope.RefScoped ?
+                    _localScopeDepth :
+                    Binder.CurrentMethodScope;
+                valEscapeScope = scope == DeclarationScope.ValueScoped ?
+                    _localScopeDepth :
+                    Binder.CallingMethodScope;
+            }
+
+            _localEscapeScopes ??= new Dictionary<LocalSymbol, (uint RefEscapeScope, uint ValEscapeScope)>();
+            _localEscapeScopes.Add(local, (refEscapeScope, valEscapeScope));
+        }
+
+#pragma warning disable IDE0060
+        private void RemoveLocalScopes(LocalSymbol local)
+        {
+            Debug.Assert(_localEscapeScopes is { });
+            // https://github.com/dotnet/roslyn/issues/65961: Currently, analysis may require subsequent calls
+            // to GetRefEscape(), etc. for the same expression so we cannot remove locals eagerly.
+            //_localEscapeScopes.Remove(local);
```

Comment:
If my understanding is correct, there is no visible harm to leaving locals in the dictionary once we leave a scope. Is that right?

Reply:
Correct, there shouldn't be any impact from keeping the locals (or the placeholders) in the dictionary beyond the end of the scope.